truffle/src/main/ruby/core/rubinius/common/random.rb

```diff
@@ -0,0 +1,203 @@
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+# * Redistributions in binary form must reproduce the above copyright notice
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+# * Neither the name of Rubinius nor the names of its contributors
+#   may be used to endorse or promote products derived from this software
+#   without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+# Only part of Rubinius' random.rb
+# Extracted from https://github.com/rubinius/rubinius/blob/v2.4.1/kernel/common/random.rb
+#
+class Rubinius::Randomizer
+
+  def initialize
+    self.seed = generate_seed
+  end
+
+  attr_reader :seed
+  def seed=(new_seed)
+    set_seed new_seed
+    @seed = new_seed
+  end
+
+  def set_seed(new_seed)
+    Rubinius.primitive :randomizer_seed
+    raise PrimitiveFailure, "Randomizer#seed primitive failed"
+  end
+
+  def swap_seed(new_seed)
+    old_seed  = self.seed
+    self.seed = new_seed
+    old_seed
+  end
+
+  def random(limit)
+    if undefined.equal?(limit)
+      random_float
+    else
+      if limit.kind_of?(Range)
+        if time_value?(limit.min)
+          random_time_range(limit)
+        else
+          random_range(limit)
+        end
+      elsif limit.kind_of?(Float)
+        raise ArgumentError, "invalid argument - #{limit}" if limit <= 0
+        random_float * limit
+      else
+        limit_int = Rubinius::Type.coerce_to limit, Integer, :to_int
+        raise ArgumentError, "invalid argument - #{limit}" if limit_int <= 0
+
+        if limit.is_a?(Integer)
+          random_integer(limit - 1)
+        elsif limit.respond_to?(:to_f)
+          random_float * limit
+        else
+          random_integer(limit_int - 1)
+        end
+      end
+    end
+  end
+
+  # Generate a random Float, in the range 0...1.0
+  def random_float
+    Rubinius.primitive :randomizer_rand_float
+    raise PrimitiveFailure, "Randomizer#rand_float primitive failed"
+  end
+
+  # Generate a random Integer, in the range 0...limit
+  def random_integer(limit)
```

Comment:
Yeah, this is a subtle bit of the semantics of Rubinius. `Rubinius.primitive :randomizer_rand_int` is a bit like a `super` call, in that it implicitly sends the same arguments as the enclosing method had. So it's as if you had written `Rubinius.primitive :randomizer_rand_int(limit)` (but you can't actually write that). That's how the `limit` argument gets into the primitive.